api/v1beta2/route_selectors_test.go

```diff
@@ -0,0 +1,211 @@
+								Type:  &[]gatewayapiv1beta1.PathMatchType{gatewayapiv1beta1.PathMatchPathPrefix}[0],
+								Value: &[]string{"/toy"}[0],
+							},
+							Method: &[]gatewayapiv1beta1.HTTPMethod{gatewayapiv1beta1.HTTPMethod("POST")}[0],
+						},
+					},
+				},
+				{
+					Matches: []gatewayapiv1beta1.HTTPRouteMatch{
+						// /assets*
+						{
+							Path: &gatewayapiv1beta1.HTTPPathMatch{
+								Type:  &[]gatewayapiv1beta1.PathMatchType{gatewayapiv1beta1.PathMatchPathPrefix}[0],
+								Value: &[]string{"/assets"}[0],
+							},
+						},
+					},
+				},
+			},
+		},
+	}
+
+	testCases := []struct {
+		name          string
+		routeSelector RouteSelector
+		route         *gatewayapiv1beta1.HTTPRoute
+		expected      []gatewayapiv1beta1.HTTPRouteRule
+	}{
+		{
+			name:          "empty route selector selects all HTTPRouteRules",
+			routeSelector: RouteSelector{},
+			route:         route,
+			expected:      route.Spec.Rules,
+		},
+		{
+			name: "route selector selects the HTTPRouteRules whose set of HTTPRouteMatch is a perfect match",
+			routeSelector: RouteSelector{
+				Matches: []gatewayapiv1beta1.HTTPRouteMatch{
+					{
+						Path: &gatewayapiv1beta1.HTTPPathMatch{
+							Type:  &[]gatewayapiv1beta1.PathMatchType{gatewayapiv1beta1.PathMatchPathPrefix}[0],
+							Value: &[]string{"/assets"}[0],
+						},
+					},
+				},
+			},
+			route:    route,
+			expected: []gatewayapiv1beta1.HTTPRouteRule{route.Spec.Rules[1]},
+		},
+		{
+			name: "route selector selects the HTTPRouteRules whose set of HTTPRouteMatch contains at least one match",
+			routeSelector: RouteSelector{
+				Matches: []gatewayapiv1beta1.HTTPRouteMatch{
+					{
+						Path: &gatewayapiv1beta1.HTTPPathMatch{
+							Type:  &[]gatewayapiv1beta1.PathMatchType{gatewayapiv1beta1.PathMatchPathPrefix}[0],
+							Value: &[]string{"/toy"}[0],
+						},
+						Method: &[]gatewayapiv1beta1.HTTPMethod{gatewayapiv1beta1.HTTPMethod("POST")}[0],
+					},
+				},
+			},
+			route:    route,
+			expected: []gatewayapiv1beta1.HTTPRouteRule{route.Spec.Rules[0]},
+		},
+		{
+			name: "route selector with missing part of a HTTPRouteMatch still selects the HTTPRouteRules that match",
+			routeSelector: RouteSelector{
+				Matches: []gatewayapiv1beta1.HTTPRouteMatch{
+					{
+						Path: &gatewayapiv1beta1.HTTPPathMatch{
+							Type:  &[]gatewayapiv1beta1.PathMatchType{gatewayapiv1beta1.PathMatchPathPrefix}[0],
+							Value: &[]string{"/toy"}[0],
+						},
+					},
+				},
+			},
+			route:    route,
+			expected: []gatewayapiv1beta1.HTTPRouteRule{route.Spec.Rules[0]},
+		},
+		{
+			name: "route selector selects no HTTPRouteRule when no criterion matches",
+			routeSelector: RouteSelector{
+				Matches: []gatewayapiv1beta1.HTTPRouteMatch{
+					{
+						Path: &gatewayapiv1beta1.HTTPPathMatch{
+							Type:  &[]gatewayapiv1beta1.PathMatchType{gatewayapiv1beta1.PathMatchExact}[0],
+							Value: &[]string{"/toy"}[0],
+						},
+					},
+				},
+			},
+			route:    route,
+			expected: nil,
```

Comment:
If the route says `/toys*`, then the route selector `/toy` should match, shouldn't it?

Reply:
It should not.

Reply:
We should document this very clearly. For me it is counter intuitive.

Reply:
`/toys*` should not include `/toy`, but `/toy*` would

Reply:
Neither `/toy`, nor `/toy*` in a route selector match selects a HTTPRouteRule that specifies a single HTTPRouteMatch `/toys*`.

For a route selector to select a HTTPRouteRule that specifies a single HTTPRouteMatch `/toys*`, the route selector MUST specify `/toys*`. I.e. the fields of route selector's HTTPRouteMatch must be identical to the same fields in HTTPRouteRule's HTTPRouteMatch. Otherwise, we're back to having to deal with the sets and subsets problem all over again.

<br/>

Here's the semantics to have in mind for each pair of route selector HTTPRouteMatch and HTTPRouteRule's HTTPRouteMatch to decide whether the route selector selects the HTTPRouteRule or not:

1. A route selector selects a HTTPRouteRule if the HTTPRouteRule contains at least one HTTPRouteMatch that specifies fields that are literally identical to all the fields specified by at least one HTTPRouteMatch of the route selector.
2. A HTTPRouteMatch within a HTTPRouteRule may include other fields that are not specified in a route selector match and yet the route selector match selects the HTTPRouteRule if rule all fields of the route select match are identically included in the HTTPRouteRule's HTTPRouteMatch; the opposite is NOT true.
3. Each field `path` of a HTTPRouteMatch, as well as each field `method` of a HTTPRouteMatch, as well as each element of the fields `headers` and `queryParams` of a HTTPRouteMatch, is atomic – this is true for the HTTPRouteMatches within a HTTPRouteRule, as well as for HTTPRouteMatches of a route selector.

This semantics is described through examples in [RFC 0001](https://github.com/Kuadrant/architecture/blob/main/rfcs/0001-rlp-v2.md).